File: wheel/src/main/java/dev/yong/wheel/utils/MD5.java
package dev.yong.wheel.utils;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * @author coderyong
 */
public final class MD5 {

    private MD5() {
    }

    private static final char[] hexDigits =
            {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};

    private static String toHexString(byte[] bytes) {
        if (bytes == null) return "";
        StringBuilder hex = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            hex.append(hexDigits[(b >> 4) & 0x0F]);
            hex.append(hexDigits[b & 0x0F]);
        }
        return hex.toString();
    }

    /**
     * 加密文件
     *
     * @param file 待加密文件
     * @return MD5加密字符串
     * @throws IOException IO操作异常
     */
    public static String encrypt(File file) throws IOException {
        MessageDigest messagedigest;
        FileInputStream in;
        FileChannel ch;
        byte[] encodeBytes;
        try {
            messagedigest = MessageDigest.getInstance("MD5");
            in = new FileInputStream(file);
            ch = in.getChannel();
            MappedByteBuffer byteBuffer = ch.map(FileChannel.MapMode.READ_ONLY, 0, file.length());
            messagedigest.update(byteBuffer);
            encodeBytes = messagedigest.digest();
        } catch (NoSuchAlgorithmException neverHappened) {
            throw new RuntimeException(neverHappened);
        }
        return toHexString(encodeBytes);
    }

    /**
     * 加密字符串
     *
     * @param str 待加密串
     * @return MD5加密字符串
     */
    public static String encrypt(String str) {
        byte[] encodeBytes;
        try {
            encodeBytes = MessageDigest.getInstance("MD5").digest(str.getBytes());
        } catch (NoSuchAlgorithmException neverHappened) {
            throw new RuntimeException(neverHappened);
        }
        return toHexString(encodeBytes);
    }
}

File: wheel/src/main/java/dev/yong/wheel/utils/MD5.kt
@file:Suppress("unused")

package dev.yong.wheel.utils

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.channels.FileChannel
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

/**
 * @author coderyong
 */
object MD5 {

    private val hexDigits =
        charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')

    private fun toHexString(bytes: ByteArray?): String {
        if (bytes == null) return ""
        val hex = StringBuilder(bytes.size * 2)
        for (b in bytes) {
            hex.append(hexDigits[(b.toInt() shr 4) and 0x0F])
            hex.append(hexDigits[b.toInt() and 0x0F])
        }
        return hex.toString()
    }

    /**
     * 加密文件
     *
     * @param file 待加密文件
     * @return MD5加密字符串
     * @throws IOException IO操作异常
     */
    @JvmStatic
    @Throws(IOException::class)
    fun encrypt(file: File): String {
        val digest: MessageDigest
        val input: FileInputStream
        val ch: FileChannel
        val encodeBytes: ByteArray
        try {
            digest = MessageDigest.getInstance("MD5")
            input = FileInputStream(file)
            ch = input.channel
            val byteBuffer = ch.map(FileChannel.MapMode.READ_ONLY, 0, file.length())
            digest.update(byteBuffer)
            encodeBytes = digest.digest()
        } catch (neverHappened: NoSuchAlgorithmException) {
            throw RuntimeException(neverHappened)
        }
        return toHexString(encodeBytes)
    }

    /**
     * 加密字符串
     *
     * @param str 待加密串
     * @return MD5加密字符串
     */
    @JvmStatic
    fun encrypt(str: String): String {
        return toHexString(
            try {
                MessageDigest.getInstance("MD5").digest(str.toByteArray())
            } catch (neverHappened: NoSuchAlgorithmException) {
                throw RuntimeException(neverHappened)
            }
        )
    }
}